stats/src/main/java/io/airlift/stats/JmxGcMonitor.java

```diff
@@ -0,0 +1,276 @@
+    @GuardedBy("this")
+    private long lastGcEndTime = System.currentTimeMillis();
+
+    @PostConstruct
+    public void start()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().addNotificationListener(
+                        objectName,
+                        notificationListener,
+                        null,
+                        null);
+            }
+            catch (JMException e) {
+                throw new RuntimeException("Unable to add GC listener", e);
+            }
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            ObjectName objectName = mbean.getObjectName();
+            try {
+                ManagementFactory.getPlatformMBeanServer().removeNotificationListener(objectName, notificationListener);
+            }
+            catch (JMException ignored) {
+            }
+        }
+    }
+
+    @Override
+    public long getMajorGcCount()
+    {
+        return majorGcCount.get();
+    }
+
+    @Override
+    public Duration getMajorGcTime()
+    {
+        return new Duration(majorGcTime.get(), MILLISECONDS);
+    }
+
+    @Managed
+    @Nested
+    public TimeStat getMajorGc()
+    {
+        return majorGc;
+    }
+
+    @Managed
+    @Nested
+    public TimeStat getMinorGc()
+    {
+        return minorGc;
+    }
+
+    @Managed
+    @Nested
+    public TimeStat getApplicationTime()
+    {
+        return applicationTime;
+    }
+
+    private synchronized void onNotification(Notification notification)
+    {
+        if ("com.sun.management.gc.notification".equals(notification.getType())) {
+            GarbageCollectionNotificationInfo info = new GarbageCollectionNotificationInfo((CompositeData) notification.getUserData());
+
+            if (info.isMajorGc()) {
+                majorGcCount.incrementAndGet();
+                majorGcTime.addAndGet(info.getDurationMs());
+                majorGc.add(info.getDurationMs(), MILLISECONDS);
+
+                // assume that major GCs stop the application
+                long applicationRuntime = max(0, info.getStartTime() - lastGcEndTime);
+                applicationTime.add(applicationRuntime, MILLISECONDS);
+                lastGcEndTime = info.getEndTime();
+
+                log.info(
+                        "Major GC: application %sms, stopped %sms:: %s -> %s",
+                        applicationRuntime,
+                        info.getDurationMs(),
+                        info.getBeforeGcTotalMb(),
+                        info.getAfterGcTotalMb());
+            }
+            else if (info.isMinorGc()) {
+                minorGc.add(info.getDurationMs(), MILLISECONDS);
+
+                // assume that minor GCs run currently
```

Comment:
It's not that simple. For CMS, young generation collections are stop-the-world (it uses the parallel collector). For G1, young and mixed collections have a stop-the-world behavior (evacuation pause, etc) and can be significant if the GC is not tuned correctly (target pause, young generation size and so on).